doc/man/nsdb-describe.8
.\"@(#)nsdb-describe.8"
.\"
.\" @file doc/man/nsdb-describe.8
.\" @brief man page for nsdb-describe client command
.\"

.\"
.\" Copyright 2011 Oracle.  All rights reserved.
.\"
.\" This file is part of fedfs-utils.
.\"
.\" fedfs-utils is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License version 2.0 as
.\" published by the Free Software Foundation.
.\"
.\" fedfs-utils is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License version 2.0 for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" version 2.0 along with fedfs-utils.  If not, see:
.\"
.\"	http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt
.\"
.TH NSDB-DESCRIBE 8 "@publication-date@"
.SH NAME
nsdb-describe \- modify the fedfsDescr attribute of a FedFS NSDB record
.SH SYNOPSIS
.B nsdb-describe
.RB [ \-?dy ]
.RB [ \-a
.IR description ]
.RB [ \-D
.IR binddn ]
.RB [ \-l
.IR nsdbname ]
.RB [ \-r
.IR nsdbport ]
.I distinguished-name
.SH INTRODUCTION
RFC 5716 introduces the Federated File System (FedFS, for short).
FedFS is an extensible standardized mechanism
by which system administrators construct
a coherent namespace across multiple file servers using
.IR "file system referrals" .
For further details, see
.BR fedfs (7).
.P
The bulk of FedFS junction information in a FedFS domain is stored
on one or more LDAP servers.
These servers are known as
.IR "namespace databases" ,
or NSDBs, for short.
.P
FedFS-enabled file servers and clients access the information stored
on NSDBs via standard LDAP queries.
FedFS-enabled file servers use these queries to resolve FedFS junctions.
FedFS administrators use them to manage information
about file sets contained in a FedFS domain name space.
.SH DESCRIPTION
The
.BR nsdb-describe (8)
command is part of a collection of low-level single-use programs
that is intended for testing the NSDB protocol or for use in scripts.
It allows FedFS administrators to update the
.B fedfsDescr
attribute of FedFS records stored on an NSDB.
.P
This command has one positional parameter which specifies the
LDAP distinguished name of the FedFS record to be modified.
All FedFS object classes may have a
.B fedfsDescr
attribute, thus a fully qualified distinguished name, rather than, say,
an FSN UUID by itself, must be specified.
.P
The
.B fedfsDescr
attribute is multi-valued.
Each attribute value is an unstructured string.
These strings may contain any valid UTF-8 character.
.P
The
.BR nsdb-describe (8)
command inserts new values or deletes or replaces existing ones
while maintaining the correct structure of the
.B fedfsDescr
attribute.
.P
Each value of the
.B fedfsDescr
attribute has no meaning to FedFS and is ignored.
Adding a description allows
free-form documentation of a FedFS record to be stored with it
without requiring changes to the NSDB's FedFS schema.
.SH OPTIONS
.IP "\fB\-a, \-\-description=\fIdescription-text\fP"
Specifies a single
.B fedfsDescr
string to be added to or deleted from the attribute's value.
The form of the string is not checked by the
.BR
nsdb-describe (8)
command.
If the
.B \-\-delete
option is specified and this string exists as a value of the target record's
.B fedfsDescr
attribute, it is removed.
Otherwise the value is added.
.IP "\fB\-d, \-\-debug"
Enables debugging messages during operation.
.IP "\fB\-?, \-\-help"
Displays
.BR nsdb-describe (8)
version information and a usage message on
.IR stderr .
.IP "\fB-D, \-\-binddn=\fIbind-distinguished-name\fP"
Specifies a distinguished name of an entity used to bind to the LDAP server
where the NSDB resides.
If the
.B \-\-binddn
option is not specified,
the value of the FEDFS_NSDB_ADMIN environment variable is consulted.
If this variable is not set,
the NSDB connection parameter database is searched for this DN.
If none of these is specified, or
if this entity does not have permission to modify this area
of the server's DIT, the
.BR nsdb-describe (8)
command fails.
.IP "\fB\-l, \-\-nsdbname=\fINSDB-hostname\fP"
Specifies the hostname of the NSDB where the target record resides.
If the
.B \-\-nsdbname
option is not specified,
the value of the FEDFS_NSDB_HOST environment variable is consulted.
If the variable is not set and the
.B \-\-nsdbname
option is not specified, the
.BR nsdb-describe (8)
command fails.
.IP "\fB\-r, \-\-nsdbport=\fINSDB-port\fP"
Specifies the IP port of the NSDB where the target record resides.
If the
.B \-\-nsdbport
option is not specified,
the value of the FEDFS_NSDB_PORT environment variable is consulted.
The default value if the variable is not set is 389.
.IP "\fB\-y, \-\-delete\fP"
Specifies that the specified value string is deleted rather than added.
.SH EXIT CODES
The NSDB returns a value that reflects the success of the requested operation.
.TP
.B FEDFS_OK
The LDAP modify request succeeded.
.TP
.B FEDFS_ERR_ACCESS
The bound entity does not have permission to perform the requested operation.
.TP
.B FEDFS_ERR_INVAL
One of the arguments was not valid.
.TP
.B FEDFS_ERR_SVRFAULT
An unanticipated non-protocol error occurred.
.TP
.B FEDFS_ERR_NSDB_ROUTE
The
.BR nsdb-describe (8)
command was unable to find a route to the specified NSDB.
.TP
.B FEDFS_ERR_NSDB_DOWN
The
.BR nsdb-describe (8)
command determined that the specified NSDB was down.
.TP
.B FEDFS_ERR_NSDB_CONN
The
.BR nsdb-describe (8)
command was unable to establish a connection with the specified NSDB.
.TP
.B FEDFS_ERR_NSDB_AUTH
The
.BR nsdb-describe (8)
command was unable to authenticate
and establish a secure connection with the specified NSDB.
.TP
.B FEDFS_ERR_NSDB_LDAP
A non-specific LDAP error occurred on the connection between the
.BR nsdb-describe (8)
command and specified NSDB.
.TP
.B FEDFS_ERR_NSDB_LDAP_VAL
An LDAP error occurred on the connection between the
.BR nsdb-describe (8)
command and specified NSDB.
The specific error may be displayed on the command line.
.TP
.B FEDFS_ERR_NSDB_NONCE
The
.BR nsdb-describe (8)
command was unable to locate the NCE on the specified NSDB.
.TP
.B FEDFS_ERR_NSDB_NOFSN
The
.BR nsdb-describe (8)
command was unable to locate the specified FSN on the specified NSDB.
.TP
.B FEDFS_ERR_NSDB_NOFSL
The
.BR nsdb-describe (8)
command was unable to locate any FSLs for the specified FSN
on the specified NSDB.
.TP
.B FEDFS_ERR_NSDB_RESPONSE
The
.BR nsdb-describe (8)
command received a malformed response from the specified NSDB.
.TP
.B FEDFS_ERR_NSDB_FAULT
An unanticipated error related to the specified NSDB occurred.
.TP
.B FEDFS_ERR_NSDB_PARAMS
The local NSDB connection parameter database
does not have any connection parameters on record for the specified NSDB.
.TP
.B FEDFS_ERR_NSDB_LDAP_REFERRAL
The
.BR nsdb-describe (8)
command received an LDAP referral that it was unable to follow.
.TP
.B FEDFS_ERR_NSDB_LDAP_REFERRAL_VAL
The
.BR nsdb-describe (8)
command received an LDAP referral that it was unable to follow.
A specific error may be displayed on the command line.
.TP
.B FEDFS_ERR_NSDB_LDAP_REFERRAL_NOTFOLLOWED
The
.BR nsdb-describe (8)
command received an LDAP referral that it chose not to follow,
either because the local implementation does not support
following LDAP referrals or LDAP referral following is disabled.
.TP
.B FEDFS_ERR_NSDB_PARAMS_LDAP_REFERRAL
The
.BR nsdb-describe (8)
command received an LDAP referral that it chose not to follow
because the local NSDB connection parameter database had no
connection parameters for the NSDB targeted by the LDAP referral.
.SH EXAMPLES
Suppose you are the FedFS administrator of the
.I example.net
FedFS domain, and you want to modify the record for
FSN UUID dc25a644-06e4-11e0-ae55-000c29dc7f8a on
the NSDB
.IR nsdb.example.net .
You might use:
.RS
.sp
$ nsdb-describe -l nsdb.example.net \\
.br
	-a "Hello, world\\!" -D cn=Manager \\
.br
	fedfsFsnUuid=dc25a644-06e4-\\
.br
	11e0-ae55-000c29dc7f8a,o=fedfs
.br
Enter NSDB password:
.br
Successfully updated description value for
.br
  fedfsFsnUuid=dc25a644-06e4-11e0-ae55-000c29dc7f8a,o=fedfs
.sp
.RE
To see the new description, use
.BR nsdb-resolve-fsn (8).
.SH SECURITY
Permission to modify the LDAP's DIT is required to update an LDAP entry.
The
.BR nsdb-describe (8)
command must bind as an entity permitted to modify the DIT
to perform this operation.
The
.BR nsdb-describe (8)
command asks for a password on
.IR stdin .
Standard password blanking techniques are used
to obscure the password on the user's terminal.
.P
The target LDAP server must be registered in the local NSDB connection
parameter database.
The connection security mode listed
in the NSDB connection parameter database
for the target LDAP server is used during this operation.
See
.BR nsdbparams (8)
for details on how to register an NSDB
in the local NSDB connection parameter database.
.SH "SEE ALSO"
.BR fedfs (7),
.BR nsdb-resolve-fsn (8),
.BR nsdbparams (8)
.sp
RFC 5716 for FedFS requirements and overview
.sp
RFC 4510 for an introduction to LDAP
.SH COLOPHON
This page is part of the fedfs-utils package.
A description of the project and information about reporting bugs
can be found at
.IR http://wiki.linux-nfs.org/wiki/index.php/FedFsUtilsProject .
.SH "AUTHOR"
Chuck Lever <chuck.lever@oracle.com>


src/nsdbc/nsdb-describe.c
/**
 * @file src/nsdbc/nsdb-describe.c
 * @brief Update the FedFsDescription attribute of a FedFS entry
 */

/*
 * Copyright 2010 Oracle.  All rights reserved.
 *
 * This file is part of fedfs-utils.
 *
 * fedfs-utils is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2.0 as
 * published by the Free Software Foundation.
 *
 * fedfs-utils is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License version 2.0 for more details.
 *
 * You should have received a copy of the GNU General Public License
 * version 2.0 along with fedfs-utils.  If not, see:
 *
 *	http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt
 */

#include <sys/types.h>
#include <sys/stat.h>

#include <stdbool.h>
#include <stdint.h>
#include <stdlib.h>
#include <stdio.h>

#include <string.h>
#include <fcntl.h>
#include <unistd.h>
#include <errno.h>
#include <getopt.h>
#include <netdb.h>
#include <locale.h>
#include <langinfo.h>

#include "fedfs.h"
#include "nsdb.h"
#include "xlog.h"
#include "gpl-boiler.h"

/**
 * Short form command line options
 */
static const char nsdb_describe_opts[] = "?a:dD:l:r:y";

/**
 * Long form command line options
 */
static const struct option nsdb_describe_longopts[] = {
	{ "binddn", 1, NULL, 'D', },
	{ "debug", 0, NULL, 'd', },
	{ "delete", 0, NULL, 'y', },
	{ "description", 1, NULL, 'a', },
	{ "help", 0, NULL, '?', },
	{ "nsdbname", 1, NULL, 'l', },
	{ "nsdbport", 1, NULL, 'r', },
	{ NULL, 0, NULL, 0, },
};

/**
 * Display program synopsis
 *
 * @param progname NUL-terminated C string containing name of program
 */
static void
nsdb_describe_usage(const char *progname)
{
	fprintf(stderr, "\n%s version " VERSION "\n", progname);
	fprintf(stderr, "Usage: %s [ -d ] [ -D binddn ] "
			"[ -l nsdbname ] [ -r nsdbport ] [ -a description] "
			"distinguished-name [-y]\n\n",
			progname);

	fprintf(stderr, "\t-?, --help           Print this help\n");
	fprintf(stderr, "\t-a, --description    Description value to modify\n");
	fprintf(stderr, "\t-d, --debug          Enable debug messages\n");
	fprintf(stderr, "\t-D, --binddn         Bind DN\n");
	fprintf(stderr, "\t-l, --nsdbname       NSDB hostname\n");
	fprintf(stderr, "\t-r, --nsdbport       NSDB port\n");
	fprintf(stderr, "\t-y, --delete         Delete specified description\n");

	fprintf(stderr, "%s", fedfs_gpl_boilerplate);

	exit((int)FEDFS_ERR_INVAL);
}

/**
 * Program entry point
 *
 * @param argc count of command line arguments
 * @param argv array of NUL-terminated C strings containing command line arguments
 * @return program exit status
 */
int
main(int argc, char **argv)
{
	char *progname, *binddn, *nsdbname;
	char *description, *entry;
	unsigned short nsdbport;
	unsigned int ldap_err;
	FedFsStatus retval;
	_Bool delete;
	nsdb_t host;
	int arg;

	(void)umask(S_IRWXO);

	/* Ensure UTF-8 strings can be handled transparently */
	if (setlocale(LC_CTYPE, "") == NULL ||
	    strcmp(nl_langinfo(CODESET), "UTF-8") != 0) {
		fprintf(stderr, "Failed to set locale and langinfo\n");
		exit((int)FEDFS_ERR_INVAL);
	}

	/* Set the basename */
	if ((progname = strrchr(argv[0], '/')) != NULL)
		progname++;
	else
		progname = argv[0];

	/* For the libraries */
	xlog_stderr(1);
	xlog_syslog(0);
	xlog_open(progname);

	nsdb_env(&nsdbname, &nsdbport, &binddn, NULL);

	delete = false;
	entry = description = NULL;
	while ((arg = getopt_long(argc, argv, nsdb_describe_opts,
			nsdb_describe_longopts, NULL)) != -1) {
		switch (arg) {
		case 'a':
			description = optarg;
			break;
		case 'd':
			xlog_config(D_ALL, 1);
			break;
		case 'D':
			binddn = optarg;
			break;
		case 'l':
			nsdbname = optarg;
			break;
		case 'r':
			if (!nsdb_parse_port_string(optarg, &nsdbport)) {
				fprintf(stderr, "Bad port number: %s\n",
					optarg);
				nsdb_describe_usage(progname);
			}
			break;
		case 'y':
			delete = true;
			break;
		default:
			fprintf(stderr, "Invalid command line "
				"argument: %c\n", (char)arg);
		case '?':
			nsdb_describe_usage(progname);
		}
	}
	if (argc == optind + 1)
		entry = argv[optind];
	else if (argc > optind + 1) {
		fprintf(stderr, "Unrecognized positional parameters\n");
		nsdb_describe_usage(progname);
	} else {
		fprintf(stderr, "No distinguished name was specified\n");
		nsdb_describe_usage(progname);
	}
	if (nsdbname == NULL) {
		fprintf(stderr, "No NSDB hostname was specified\n");
		nsdb_describe_usage(progname);
	}
	if (description == NULL && !delete) {
		fprintf(stderr, "No description was specified\n");
		nsdb_describe_usage(progname);
	}

	retval = nsdb_lookup_nsdb(nsdbname, nsdbport, &host, NULL);
	switch (retval) {
	case FEDFS_OK:
		break;
	case FEDFS_ERR_NSDB_PARAMS:
		fprintf(stderr, "No connection parameters for NSDB %s:%u\n",
			nsdbname, nsdbport);
		goto out;
	default:
		fprintf(stderr, "Failed to look up NSDB %s:%u: %s\n",
			nsdbname, nsdbport,
			nsdb_display_fedfsstatus(retval));
		goto out;
	}
	retval = FEDFS_ERR_INVAL;
	if (binddn == NULL)
		binddn = (char *)nsdb_default_binddn(host);
	if (binddn == NULL) {
		fprintf(stderr, "No NDSB bind DN was specified\n");
		goto out_free;
	}

	retval = nsdb_open_nsdb(host, binddn, NULL, &ldap_err);
	switch (retval) {
	case FEDFS_OK:
		break;
	case FEDFS_ERR_NSDB_CONN:
		fprintf(stderr, "Failed to connect to NSDB %s:%u\n",
			nsdbname, nsdbport);
		goto out_free;
	case FEDFS_ERR_NSDB_AUTH:
		fprintf(stderr, "Failed to authenticate to NSDB %s:%u\n",
			nsdbname, nsdbport);
		goto out_free;
	case FEDFS_ERR_NSDB_LDAP_VAL:
		fprintf(stderr, "Failed to authenticate to NSDB %s:%u: %s\n",
			nsdbname, nsdbport, ldap_err2string(ldap_err));
		goto out_free;
	default:
		fprintf(stderr, "Failed to bind to NSDB %s:%u: %s\n",
			nsdbname, nsdbport,
			nsdb_display_fedfsstatus(retval));
		goto out_free;
	}

	if (delete)
		retval = nsdb_description_delete_s(host, entry,
							description, &ldap_err);
	else
		retval = nsdb_description_add_s(host, entry,
							description, &ldap_err);
	switch (retval) {
	case FEDFS_OK:
		printf("Successfully %s description value %s %s\n",
			delete ? "removed" : "updated",
			delete ? "from" : "for", entry);
		break;
	case FEDFS_ERR_NSDB_LDAP_VAL:
		switch (ldap_err) {
		case LDAP_NO_SUCH_OBJECT:
			fprintf(stderr, "Entry \"%s\" not found\n", entry);
			break;
		case LDAP_NO_SUCH_ATTRIBUTE:
			fprintf(stderr, "Description value \"%s\" not found\n",
				description);
			break;
		default:
			fprintf(stderr, "Failed to %s description value for %s: %s\n",
				delete ? "remove" : "update", entry,
				ldap_err2string(ldap_err));
		}
		break;
	default:
		fprintf(stderr, "Failed to update description value for %s: %s\n",
			entry, nsdb_display_fedfsstatus(retval));
	}

	nsdb_close_nsdb(host);

out_free:
	nsdb_free_nsdb(host);

out:
	exit((int)retval);
}
